app/models/course.rb

```diff
@@ -22,13 +22,48 @@ def notes_number_hash
     ]
   end
 
-  def first_try_accepted_tasks_number_hash
+  def accepted_tasks_number_hash
+      Hash[
+        self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.flatten.
+          select { |task| task.accepted?} .count }
+      ]
+  end
+
+  def first_try_accepted_number_hash
     Hash[
       self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.flatten.
         select { |task| task.accepted? && task.notes.count == 0 } .count }
     ]
   end
 
+  def attempts_to_pass_number_hash
+    Hash[
+      self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.
+        collect { |task| task.submissions } .flatten.count }
+    ]
+  end
+
+  def max_of_attempts_to_pass_number_hash
+    Hash[
+      self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.
+        max_by { |task| task.submissions.count } .submissions.count }
+      ]
+  end
+
+  def min_attempts_number_problem_hash
+    Hash[
+      self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.
+        min_by { |task| task.submissions.count } .problem }
+      ]
+  end
+
+  def max_attempts_number_problem_hash
+    Hash[
+      self.terms.zip Term.where(course_id: self.id).collect { |term| term.tasks.
```

Comment:
Maybe try to remove this duplication? And possibly refactor the way the data is collected.